src/main/java/uk/ac/ic/wlgitbridge/data/PostgresProjectLockImpl.java

```diff
@@ -0,0 +1,106 @@
+package uk.ac.ic.wlgitbridge.data;
+
+import org.apache.commons.dbcp2.BasicDataSource;
+import uk.ac.ic.wlgitbridge.bridge.db.postgres.PostgresRowLock;
+import uk.ac.ic.wlgitbridge.bridge.lock.ProjectLock;
+
+import java.sql.Connection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class PostgresProjectLockImpl implements ProjectLock {
+
+  // Keep locks in a map, for lookup by name
+  private final Map<String, Lock> projectLocks;
+
+  // An app-level lock that tracks all current "operations", or,
+  // threads that currently hold a lock. This allows us to wait on
+  // the wlock, effectively waiting for all outstanding operations to
+  // complete, before shutting down
+  private final ReentrantReadWriteLock rwlock;
+  private final Lock rlock;
+  private final ReentrantReadWriteLock.WriteLock wlock;
+
+  // Postgres connection pool
+  private final BasicDataSource pool;
+
+  private LockAllWaiter waiter;
+  private boolean waiting;
+
+  public PostgresProjectLockImpl(BasicDataSource connectionPool) {
+    projectLocks = new HashMap<String, Lock>();
+    rwlock = new ReentrantReadWriteLock();
+    rlock = rwlock.readLock();
+    wlock = rwlock.writeLock();
+    pool = connectionPool;
+    waiting = false;
+  }
+
+  public PostgresProjectLockImpl(BasicDataSource connectionPool, LockAllWaiter waiter) {
+    this(connectionPool);
+    setWaiter(waiter);
+  }
+
+  private PostgresRowLock getProjectLock(String projectName) {
+    try {
+      Connection connection = pool.getConnection();
+      PostgresRowLock lock = new PostgresRowLock(projectName, connection);
+      return lock;
+    } catch (Exception e) {
+      throw new RuntimeException("Postgres query error", e);
+    }
+  }
+
+  private synchronized Lock getLockForProjectName(String projectName) {
```

Comment:
I think it's my lack of experience with Java here, but I don't understand why we need this `synchronized` method to wrap `getProjecLock`. It seems both take a `projectName`. - Could we have a single implementation instead of two methods, or is there a reason to do it like this?

Do you know which part of this mechanism requires the `synchronized` keyword, here?

Reply:
So, `synchronized` means only one thread can run this code at a time. I _think_ this could be a stray addition that snuck in while debugging a concurrency problem. I'll see if it can be removed :+1:

Reply:
Ah, so we have the same `synchronized` keyword on the other implementation. 

I think _this_ is what makes the other implementation thread-safe in accessing the hash-map. But there is a problem in this new implementation, as we're accessing the hash-map outside of the sync block. I'll jiggle these around to make it safe again :+1:

Reply:
(Times like these it would be nice to have some java experts around :) )

Reply:
> Times like these it would be nice to have some java experts around

Yeah. I was always "the threads guy" when I was doing C++, which helps, but `synchronized` isn't something we have (or had, at least - I haven't kept up with the latest changes) over there.